File: g2i_route_sync/gpx.py
"""GPX construction and format detection helpers."""

from __future__ import annotations

import datetime as dt
from typing import Any
from xml.sax.saxutils import escape


def infer_extension(content: bytes, source: str) -> str:
    """Guess the route file extension from its source URL and/or content."""
    lowered = source.lower()
    if lowered.endswith(".gpx"):
        return ".gpx"
    if lowered.endswith(".tcx"):
        return ".tcx"
    if lowered.endswith(".fit"):
        return ".fit"
    if content[:100].lstrip().startswith(b"<?xml"):
        if b"<gpx" in content[:500].lower():
            return ".gpx"
        if b"<trainingcenterdatabase" in content[:1000].lower():
            return ".tcx"
    return ".fit"


def build_gpx_from_course_detail(
    course_detail: dict[str, Any], fallback_name: str
) -> bytes:
    """Build a GPX document from a Garmin course detail payload."""
    geo_points = course_detail.get("geoPoints")
    if not isinstance(geo_points, list) or not geo_points:
        raise RuntimeError("Garmin course detail does not contain geoPoints")

    course_points = course_detail.get("coursePoints")
    waypoints: list[str] = []
    if isinstance(course_points, list):
        for cp in course_points:
            if not isinstance(cp, dict):
                continue
            lat = cp.get("lat")
            lon = cp.get("lon")
            if lat is None or lon is None:
                continue
            wpt_name = str(cp.get("name") or "POI")
            wpt_note = cp.get("note")
            wpt_type = cp.get("coursePointType")

            lines = [f'  <wpt lat="{lat}" lon="{lon}">']
            ele = cp.get("elevation")
            if ele is not None:
                lines.append(f"    <ele>{ele}</ele>")
            lines.append(f"    <name>{escape(wpt_name)}</name>")
            if isinstance(wpt_note, str) and wpt_note.strip():
                lines.append(f"    <cmt>{escape(wpt_note)}</cmt>")
            if isinstance(wpt_type, str) and wpt_type.strip():
                lines.append(f"    <type>{escape(wpt_type)}</type>")
            lines.append("  </wpt>")
            waypoints.append("\n".join(lines))

    name = str(course_detail.get("courseName") or fallback_name)
    trkpts: list[str] = []
    for point in geo_points:
        if not isinstance(point, dict):
            continue

        lat = point.get("latitude")
        lon = point.get("longitude")
        if lat is None or lon is None:
            continue

        line_parts = [f'<trkpt lat="{lat}" lon="{lon}">']
        ele = point.get("elevation")
        if ele is not None:
            line_parts.append(f"<ele>{ele}</ele>")

        timestamp_ms = point.get("timestamp")
        if isinstance(timestamp_ms, (int, float)) and timestamp_ms > 0:
            timestamp = dt.datetime.fromtimestamp(
                float(timestamp_ms) / 1000.0, tz=dt.timezone.utc
            )
            line_parts.append(
                f"<time>{timestamp.isoformat().replace('+00:00', 'Z')}</time>"
            )

        line_parts.append("</trkpt>")
        trkpts.append("".join(line_parts))

    if not trkpts:
        raise RuntimeError("No valid geoPoints to build GPX")

    gpx = (
        '<?xml version="1.0" encoding="UTF-8"?>\n'
        '<gpx version="1.1" creator="g2i-route-sync" '
        'xmlns="http://www.topografix.com/GPX/1/1">\n'
        + ("\n".join(waypoints) + "\n" if waypoints else "")
        + f"  <trk><name>{escape(name)}</name><trkseg>\n"
        + "\n".join(f"    {trackpoint}" for trackpoint in trkpts)
        + "\n  </trkseg></trk>\n"
        + "</gpx>\n"
    )
    return gpx.encode("utf-8")

File: lib/gpx.dart
/// GPX construction and format detection helpers.
library;

import 'dart:convert';

/// Escape the XML special characters that ``xml.sax.saxutils.escape`` handles
/// by default: ``&``, ``<`` and ``>``.
String _xmlEscape(String value) => value
    .replaceAll('&', '&amp;')
    .replaceAll('<', '&lt;')
    .replaceAll('>', '&gt;');

/// Format a JSON number the way Python's ``str()`` would for GPX coordinates,
/// so doubles like ``35.0`` keep their trailing ``.0``.
String _numStr(dynamic value) {
  if (value is int) return value.toString();
  if (value is double) {
    if (value.isFinite && value == value.truncateToDouble()) {
      return '${value.toInt()}.0';
    }
    return value.toString();
  }
  return value.toString();
}

/// ISO-8601 UTC timestamp matching Python's ``datetime.isoformat()`` output:
/// fractional seconds are omitted when zero, and the offset is rendered ``Z``.
String _isoUtc(int millis) {
  final dt = DateTime.fromMillisecondsSinceEpoch(millis, isUtc: true);
  var iso = dt.toIso8601String(); // e.g. 2023-11-14T22:13:20.000Z
  iso = iso.replaceAll('.000Z', 'Z');
  return iso;
}

/// Guess the route file extension from its source URL and/or content.
String inferExtension(List<int> content, String source) {
  final lowered = source.toLowerCase();
  if (lowered.endsWith('.gpx')) return '.gpx';
  if (lowered.endsWith('.tcx')) return '.tcx';
  if (lowered.endsWith('.fit')) return '.fit';

  final head = latin1.decode(content.take(1000).toList(), allowInvalid: true);
  if (head.trimLeft().startsWith('<?xml')) {
    final first500 = head.length > 500 ? head.substring(0, 500) : head;
    if (first500.toLowerCase().contains('<gpx')) return '.gpx';
    if (head.toLowerCase().contains('<trainingcenterdatabase')) return '.tcx';
  }
  return '.fit';
}

/// Build a GPX document (UTF-8 bytes) from a Garmin course detail payload.
List<int> buildGpxFromCourseDetail(
  Map<String, dynamic> courseDetail,
  String fallbackName,
) {
  final geoPoints = courseDetail['geoPoints'];
  if (geoPoints is! List || geoPoints.isEmpty) {
    throw StateError('Garmin course detail does not contain geoPoints');
  }

  final coursePoints = courseDetail['coursePoints'];
  final waypoints = <String>[];
  if (coursePoints is List) {
    for (final cp in coursePoints) {
      if (cp is! Map) continue;
      final lat = cp['lat'];
      final lon = cp['lon'];
      if (lat == null || lon == null) continue;
      final wptName = (cp['name'] ?? 'POI').toString();
      final wptNote = cp['note'];
      final wptType = cp['coursePointType'];

      final lines = <String>[
        '  <wpt lat="${_numStr(lat)}" lon="${_numStr(lon)}">',
      ];
      final ele = cp['elevation'];
      if (ele != null) lines.add('    <ele>${_numStr(ele)}</ele>');
      lines.add('    <name>${_xmlEscape(wptName)}</name>');
      if (wptNote is String && wptNote.trim().isNotEmpty) {
        lines.add('    <cmt>${_xmlEscape(wptNote)}</cmt>');
      }
      if (wptType is String && wptType.trim().isNotEmpty) {
        lines.add('    <type>${_xmlEscape(wptType)}</type>');
      }
      lines.add('  </wpt>');
      waypoints.add(lines.join('\n'));
    }
  }

  final name = (courseDetail['courseName'] ?? fallbackName).toString();
  final trkpts = <String>[];
  for (final point in geoPoints) {
    if (point is! Map) continue;

    final lat = point['latitude'];
    final lon = point['longitude'];
    if (lat == null || lon == null) continue;

    final parts = <String>[
      '<trkpt lat="${_numStr(lat)}" lon="${_numStr(lon)}">',
    ];
    final ele = point['elevation'];
    if (ele != null) parts.add('<ele>${_numStr(ele)}</ele>');

    final timestampMs = point['timestamp'];
    if (timestampMs is num && timestampMs > 0) {
      parts.add('<time>${_isoUtc(timestampMs.toInt())}</time>');
    }

    parts.add('</trkpt>');
    trkpts.add(parts.join());
  }

  if (trkpts.isEmpty) {
    throw StateError('No valid geoPoints to build GPX');
  }

  final buffer =
      StringBuffer()
        ..write('<?xml version="1.0" encoding="UTF-8"?>\n')
        ..write(
          '<gpx version="1.1" creator="g2i-route-sync" '
          'xmlns="http://www.topografix.com/GPX/1/1">\n',
        );
  if (waypoints.isNotEmpty) {
    buffer.write('${waypoints.join('\n')}\n');
  }
  buffer
    ..write('  <trk><name>${_xmlEscape(name)}</name><trkseg>\n')
    ..write(trkpts.map((t) => '    $t').join('\n'))
    ..write('\n  </trkseg></trk>\n')
    ..write('</gpx>\n');

  return utf8.encode(buffer.toString());
}
